File: src/validate.ts
import type { Capability, Isolation, RunInput } from "./types.js";

const CAPABILITIES: Capability[] = [
  "ask",
  "plan",
  "write",
  "write-unsandboxed",
];

function invalid(field: string, detail?: string): never {
  throw new Error(
    detail ? `invalid ${field}: ${detail}` : `invalid ${field}`,
  );
}

function parseIsolation(v: unknown): Isolation {
  if (!v || typeof v !== "object") invalid("isolation");
  const o = v as Record<string, unknown>;
  switch (o.type) {
    case "None":
      return { type: "None" };
    case "CallerProvided": {
      if (typeof o.path !== "string" || o.path.length === 0) {
        invalid("isolation.path", "required non-empty string");
      }
      return { type: "CallerProvided", path: o.path };
    }
    case "BackendProvided": {
      const iso: Extract<Isolation, { type: "BackendProvided" }> = {
        type: "BackendProvided",
      };
      if (o.name !== undefined) {
        if (typeof o.name !== "string") invalid("isolation.name");
        iso.name = o.name;
      }
      if (o.base !== undefined) {
        if (typeof o.base !== "string") invalid("isolation.base");
        iso.base = o.base;
      }
      return iso;
    }
    default:
      invalid("isolation.type");
  }
}

/** Validate MCP cursor_run args into a typed RunInput. */
export function validateRunInput(args: Record<string, unknown>): RunInput {
  const prompt = args.prompt;
  if (typeof prompt !== "string" || prompt.length === 0) {
    throw new Error('missing required field "prompt"');
  }

  const input: RunInput = { prompt };

  if (args.model !== undefined) {
    if (typeof args.model !== "string" || args.model.length === 0) {
      invalid("model");
    }
    input.model = args.model;
  }

  if (args.requireNonClaude !== undefined) {
    if (typeof args.requireNonClaude !== "boolean") {
      invalid("requireNonClaude", "must be a boolean");
    }
    input.requireNonClaude = args.requireNonClaude;
  }

  if (args.capability !== undefined) {
    if (
      typeof args.capability !== "string" ||
      !CAPABILITIES.includes(args.capability as Capability)
    ) {
      invalid("capability");
    }
    input.capability = args.capability as Capability;
  }

  if (args.allowUnsandboxed !== undefined) {
    if (typeof args.allowUnsandboxed !== "boolean") {
      invalid("allowUnsandboxed", "must be a boolean");
    }
    input.allowUnsandboxed = args.allowUnsandboxed;
  }

  if (args.session !== undefined) {
    if (typeof args.session !== "string" || args.session.length === 0) {
      invalid("session");
    }
    input.session = args.session;
  }

  if (args.isolation !== undefined) {
    input.isolation = parseIsolation(args.isolation);
  }

  if (args.verifyCommands !== undefined) {
    if (!Array.isArray(args.verifyCommands)) invalid("verifyCommands");
    const cmds = args.verifyCommands.filter(
      (x): x is string => typeof x === "string",
    );
    if (cmds.length !== args.verifyCommands.length) {
      invalid("verifyCommands", "must be an array of strings");
    }
    input.verifyCommands = cmds;
  }

  if (args.gate !== undefined) {
    if (typeof args.gate !== "string") invalid("gate");
    input.gate = args.gate;
  }

  if (args.allowPartialCommit !== undefined) {
    if (typeof args.allowPartialCommit !== "boolean") {
      invalid("allowPartialCommit", "must be a boolean");
    }
    input.allowPartialCommit = args.allowPartialCommit;
  }

  if (args.waitMs !== undefined) {
    if (typeof args.waitMs !== "number" || !Number.isFinite(args.waitMs)) {
      invalid("waitMs", "must be a finite number");
    }
    input.waitMs = args.waitMs;
  }

  if (args.idleMs !== undefined) {
    if (
      args.idleMs !== null &&
      (typeof args.idleMs !== "number" || !Number.isFinite(args.idleMs))
    ) {
      invalid("idleMs", "must be a finite number or null");
    }
    input.idleMs = args.idleMs as number | null;
  }

  if (args.toolIdleMs !== undefined) {
    if (
      args.toolIdleMs !== null &&
      (typeof args.toolIdleMs !== "number" || !Number.isFinite(args.toolIdleMs))
    ) {
      invalid("toolIdleMs", "must be a finite number or null");
    }
    input.toolIdleMs = args.toolIdleMs as number | null;
  }

  if (args.background !== undefined) {
    if (typeof args.background !== "boolean") {
      invalid("background", "must be a boolean");
    }
    input.background = args.background;
  }

  return input;
}

File: src/validate.rs
use crate::types::RunInput;
use serde_json::Value;

fn invalid(field: &str, detail: Option<&str>) -> String {
    match detail {
        Some(d) => format!("invalid {field}: {d}"),
        None => format!("invalid {field}"),
    }
}

fn parse_isolation(v: &Value) -> Result<crate::types::Isolation, String> {
    let o = v.as_object().ok_or_else(|| invalid("isolation", None))?;
    let ty = o.get("type").and_then(|x| x.as_str()).unwrap_or("");
    match ty {
        "None" => Ok(crate::types::Isolation::None),
        "CallerProvided" => {
            let path = o.get("path").and_then(|x| x.as_str()).unwrap_or("");
            if path.is_empty() {
                return Err(invalid("isolation.path", Some("required non-empty string")));
            }
            Ok(crate::types::Isolation::CallerProvided {
                path: path.to_string(),
            })
        }
        "BackendProvided" => {
            let mut name = None;
            let mut base = None;
            if let Some(n) = o.get("name") {
                let s = n.as_str().ok_or_else(|| invalid("isolation.name", None))?;
                name = Some(s.to_string());
            }
            if let Some(b) = o.get("base") {
                let s = b.as_str().ok_or_else(|| invalid("isolation.base", None))?;
                base = Some(s.to_string());
            }
            Ok(crate::types::Isolation::BackendProvided { name, base })
        }
        _ => Err(invalid("isolation.type", None)),
    }
}

pub fn validate_run_input(args: &Value) -> Result<RunInput, String> {
    let obj = args.as_object();
    let prompt = obj.and_then(|o| o.get("prompt")).and_then(|p| p.as_str());
    let prompt = match prompt {
        Some(p) if !p.is_empty() => p.to_string(),
        _ => return Err("missing required field \"prompt\"".into()),
    };
    let mut input = RunInput {
        prompt,
        ..RunInput::default()
    };
    let Some(obj) = obj else {
        return Ok(input);
    };

    if let Some(v) = obj.get("model") {
        let s = v
            .as_str()
            .filter(|s| !s.is_empty())
            .ok_or_else(|| invalid("model", None))?;
        input.model = Some(s.to_string());
    }
    if let Some(v) = obj.get("requireNonClaude") {
        let b = v
            .as_bool()
            .ok_or_else(|| invalid("requireNonClaude", Some("must be a boolean")))?;
        input.require_non_claude = Some(b);
    }
    if let Some(v) = obj.get("capability") {
        let s = v.as_str().ok_or_else(|| invalid("capability", None))?;
        let cap = crate::types::Capability::parse(s).ok_or_else(|| invalid("capability", None))?;
        input.capability = Some(cap);
    }
    if let Some(v) = obj.get("allowUnsandboxed") {
        let b = v
            .as_bool()
            .ok_or_else(|| invalid("allowUnsandboxed", Some("must be a boolean")))?;
        input.allow_unsandboxed = Some(b);
    }
    if let Some(v) = obj.get("session") {
        let s = v
            .as_str()
            .filter(|s| !s.is_empty())
            .ok_or_else(|| invalid("session", None))?;
        input.session = Some(s.to_string());
    }
    if let Some(v) = obj.get("isolation") {
        input.isolation = Some(parse_isolation(v)?);
    }
    if let Some(v) = obj.get("verifyCommands") {
        let arr = v
            .as_array()
            .ok_or_else(|| invalid("verifyCommands", None))?;
        let cmds: Vec<String> = arr
            .iter()
            .filter_map(|x| x.as_str().map(|s| s.to_string()))
            .collect();
        if cmds.len() != arr.len() {
            return Err(invalid(
                "verifyCommands",
                Some("must be an array of strings"),
            ));
        }
        input.verify_commands = Some(cmds);
    }
    if let Some(v) = obj.get("gate") {
        let s = v.as_str().ok_or_else(|| invalid("gate", None))?;
        input.gate = Some(s.to_string());
    }
    if let Some(v) = obj.get("allowPartialCommit") {
        let b = v
            .as_bool()
            .ok_or_else(|| invalid("allowPartialCommit", Some("must be a boolean")))?;
        input.allow_partial_commit = Some(b);
    }
    if let Some(v) = obj.get("waitMs") {
        let n = v
            .as_f64()
            .filter(|n| n.is_finite())
            .ok_or_else(|| invalid("waitMs", Some("must be a finite number")))?;
        input.wait_ms = Some(n);
    }
    if let Some(v) = obj.get("idleMs") {
        if v.is_null() {
            input.idle_ms = Some(None);
        } else {
            let n = v
                .as_f64()
                .filter(|n| n.is_finite())
                .ok_or_else(|| invalid("idleMs", Some("must be a finite number or null")))?;
            input.idle_ms = Some(Some(n));
        }
    }
    if let Some(v) = obj.get("toolIdleMs") {
        if v.is_null() {
            input.tool_idle_ms = Some(None);
        } else {
            let n = v
                .as_f64()
                .filter(|n| n.is_finite())
                .ok_or_else(|| invalid("toolIdleMs", Some("must be a finite number or null")))?;
            input.tool_idle_ms = Some(Some(n));
        }
    }
    if let Some(v) = obj.get("background") {
        let b = v
            .as_bool()
            .ok_or_else(|| invalid("background", Some("must be a boolean")))?;
        input.background = Some(b);
    }
    Ok(input)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn accepts_model_and_require_non_claude() {
        let input = validate_run_input(&json!({
            "prompt": "do it",
            "model": "grok-4.5-xhigh",
            "requireNonClaude": true,
        }))
        .unwrap();
        assert_eq!(input.prompt, "do it");
        assert_eq!(input.model.as_deref(), Some("grok-4.5-xhigh"));
        assert_eq!(input.require_non_claude, Some(true));
    }

    #[test]
    fn rejects_empty_model() {
        let e = validate_run_input(&json!({"prompt": "x", "model": ""})).unwrap_err();
        assert!(e.contains("model"));
    }

    #[test]
    fn rejects_non_boolean_require_non_claude() {
        let e = validate_run_input(&json!({"prompt": "x", "requireNonClaude": "yes"})).unwrap_err();
        assert!(e.contains("requireNonClaude"));
    }

    #[test]
    fn tier_is_silently_ignored() {
        let input = validate_run_input(&json!({"prompt": "x", "tier": "cheap-bulk"})).unwrap();
        assert_eq!(input.prompt, "x");
        assert!(input.model.is_none());
    }

    #[test]
    fn omitted_require_non_claude_stays_none() {
        let input = validate_run_input(&json!({"prompt": "x"})).unwrap();
        assert!(input.require_non_claude.is_none());
    }
}
